### DISCUSSION TITLE
Using GridSampler as sampler in tio.Queue()

### DISCUSSION THREAD

**Discussion by MirenLurBarquin** [TorchIO-project/torchio#1061] [Q&A]
Hi all, 

in the example for `tio.data.Queue()` given in TorchIOs [documentation](https://torchio.readthedocs.io/patches/patch_training.html) the patch sampler used is `tio.data.UniformSampler`, which does not need the subject as input. Thus, I can pass a ` tio.SubjectDataset()` with multiple elements as an argument in `tio.data.Queue()` and the patches will be loaded without any problem later with `torch.utils.data.DataLoader()`.

Say that in `tio.data.Queue()` I want to use `tio.data.GridSampler()` which needs a subject as an argument. In my case I have customized the `GridSampler()` function so that it only takes patches with no background; thus, the output of `CustomGridSampler()` will be different for each subject. 

Is there any way that I can use my customized grid sampler following the given example code: 
```
patches_queue = tio.Queue(
    subjects_dataset,
    queue_length,
    samples_per_volume,
    patch_sampler,
    num_workers=4,
    subject_sampler=subject_sampler,
)
```
so that it takes the each subject as argument? 

Thank you in advance :)

**Comment by fepegar** [comment]
Hi, @MirenLurBarquin. I recommend you use a weighted sampler or a label sampler for this.

If you really want to use something like a grid sampler, you could precompute the number of patches it will take from each volume and add specify it when instantiating the `Subject`, as explained in the docs for the `samples_per_volume` parameter of the Queue.

**Reply by MirenLurBarquin** [reply]
Hi, @fepegar. First, thank you a lot for the quick response. And now I see why a weighted sampler is a feasible idea :) 

Nevertheless, I will try to explain a bit better my goal to leave here my idea of creating a custom grid sampler. I was able to fix my problem by combining the idea of `WeightedSampler` `probability_map` argument and the `GridSampler` idea. 

I am trying to do axon segmentation. Currently, I have just one subject as dataset, a `.nii` image. My sample is a cylindric volume, therefore, all the remaining voxels in the image outside from the cylinder are background and have no information. I have stored in my `tio.Subject(img=img, axons=axons_seg, cylinder=mask)` a mask for the cylindric shape.   I want to do data augmentation, therefore I make copies of the subject and transform each of them generating a bigger dataset:  

`dataset = tio.SubjectsDataset(subjects, transform=transform)`

After transforming, each subject will have a different transformed cylindric shape. In the transform is also important to include scaling.

  If I use `tio.data.WeightedSampler(patch_size, ‘cylinder’)`, I may get patches with very little information due to the scaling of some volumes. On the other hand, my `CustomGridSampler` would take just patches that fall completely into the (transformed) cylindric shape (being aware that this is not completely necessary, and that having some background should not be that big problem).  

I will try both samplers to check if there is significant difference in performance. Thank you one more time!

**Reply by romainVala** [reply]
Hi
I think `WeightedSample` would be fine since the cylinder (mask) will also be transformed (and scaled) ...
